### PR TITLE
Improve the Mixer perf testing framework.

### PR DESCRIPTION
**Pull request by ozevren**
- Add support for writing tests directly against runtime.Dispatcher interface.
- Make the perf test execution type (i.e inprocess/coprocess/dispatcher) a flag.
- Rename perf.Env to perf.Settings, both to avoid conflict with adapter.Env, and to put additional, generic settings.

### CODE REVIEW

**Comment by ozevren** [comment]
This depends on: https://github.com/istio/istio/pull/2323 Also renames dispatcherForTesting => dispatcher.

File: mixer/pkg/perf/config.go [hunk 1]
```diff
@@ -14,12 +14,16 @@
 
 package perf
 
+// TODO: We should ideally combine this file with pkg/server/Args. Unfortunately, pkg/serverArgs is not serializable.
```
Comment (geeknoid) [inline mixer/pkg/perf/config.go]: What does it take to make it serializable? Just adding the annotations?

File: mixer/pkg/perf/run.go [hunk 2]
```diff
@@ -57,22 +62,24 @@ func (b *testingBWrapper) n() int {
...
-func RunCoprocess(b *testing.B, setup *Setup, env *Env, executablePathSuffix string) {
-	run(&testingBWrapper{b: b}, setup, env, executablePathSuffix, true)
+// Run executes the benchmark using specified setup and settings.
+func Run(b *testing.B, setup *Setup, settings Settings) {
+	switch settings.RunMode {
+	case InProcess:
+		run(&testingBWrapper{b}, setup, &settings, false)
+	case CoProcess:
+		run(&testingBWrapper{b}, setup, &settings, true)
+	case DispatcherOnly:
```
Comment (geeknoid) [inline mixer/pkg/perf/run.go]: Can we perhaps make this a bit more generic? At least in its name?

Specifically, the interesting bit here is "local with no RPC overhead" right? That is, the point is not to interface with the dispatcher directly, the point is about bypassing gRPC.

File: mixer/pkg/perf/run.go [hunk 3]
```diff
@@ -130,3 +141,59 @@ func run(b benchmark, setup *Setup, env *Env, executablePathSuffix string, copro
...
+	}
+
+	s, err := mixer.New(a)
+	if err != nil {
+		b.fatalf("error creating Mixer server: '%s'", err.Error())
+		return
+	}
+	defer s.Close()
+
+	dispatcher := s.GetDispatcher()
```
Comment (geeknoid) [inline mixer/pkg/perf/run.go]: Go conventions say no "Get". So the function should just be Dispatcher()

File: mixer/pkg/perf/run.go [hunk 4]
```diff
@@ -130,3 +141,59 @@ func run(b benchmark, setup *Setup, env *Env, executablePathSuffix string, copro
...
+	requests := setup.Load.createRequestProtos(setup.Config)
+	bags := make([]attribute.Bag, len(requests)) // precreate bags to avoid polluting allocation data.
+	for i, r := range requests {
+		switch r.(type) {
+		case *istio_mixer_v1.ReportRequest:
+			rrep := r.(*istio_mixer_v1.ReportRequest)
+			bags[i] = attribute.NewProtoBag(&rrep.Attributes[0], globalDict, attribute.GlobalList())
+
+		case *istio_mixer_v1.CheckRequest:
+			crep := r.(*istio_mixer_v1.CheckRequest)
```
Comment (geeknoid) [inline mixer/pkg/perf/run.go]: Let's use a more PC-appropriate term :-)

File: mixer/pkg/perf/run.go [hunk 5]
```diff
@@ -130,3 +141,59 @@ func run(b benchmark, setup *Setup, env *Env, executablePathSuffix string, copro
...
+	}
+
+	b.run("DispatcherOnly", func(bb benchmark) {
+		for i := 0; i < bb.n(); i++ {
+			for j, r := range requests {
+				bag := bags[j]
+
+				switch r.(type) {
+				case *istio_mixer_v1.ReportRequest:
+					_ = dispatcher.Report(context.Background(), bag)
```
Comment (geeknoid) [inline mixer/pkg/perf/run.go]: Don't you want to track errors? You don't want unreported errors skewing benchmark numbers...

File: mixer/pkg/perf/settings.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+	// InProcess run mode indicates that an in-process client should be used to drive the test.
+	// This potentially affects collected profile data, as it will include client's execution as well.
+	InProcess RunMode = iota
+
+	// CoProcess run mode indicates that the client should be run in a separate process.
+	// This avoids client's execution polluting the profile data, but may cause variations in timings
+	// as the communication overhead between the client and server is counted in timings.
+	CoProcess
+
+	// DispatcherOnly run mode indicates that the test should be run against the runtime.DispatcherOnly interface
```
Comment (geeknoid) [inline mixer/pkg/perf/settings.go]: runtime.DispatcherOnly -> runtime.Dispatcher

File: mixer/pkg/perf/settings.go [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+	DispatcherOnly
+)
+
+// Settings is the perf test settings.
+type Settings struct {
+	RunMode   RunMode
+	Templates map[string]template.Info
+	Adapters  []adapter.InfoFn
+
+	// ExecutableSearchSuffix indicates the search prefix to use when trying to locate the co-process
```
Comment (geeknoid) [inline mixer/pkg/perf/settings.go]: search prefix -> search suffix

File: mixer/pkg/perf/config.go (hunk 1 above)
Comment (ozevren) [reply mixer/pkg/perf/config.go]: Annotations + Coming up with a strategy for serializing & reconciling adapters/templates during deserialization. They are currently kept as full-blown objects.

**Comment by ozevren** [comment]
Accomodated most feedback. PTAL.

**Pull request closed without merging**